arroyo-worker/src/connectors/mqtt/sink/mod.rs

```diff
@@ -0,0 +1,142 @@
+use arroyo_rpc::OperatorConfig;
+use arroyo_types::{Key, Record};
+use rumqttc::v5::mqttbytes::QoS;
+use rumqttc::v5::AsyncClient;
+use serde::Serialize;
+
+use crate::connectors::mqtt::{MqttConfig, MqttTable, QualityOfService, Retain, TableType};
+use crate::engine::{Context, StreamNode};
+
+#[cfg(test)]
+mod test;
+
+#[derive(StreamNode)]
+pub struct MqttSinkFunc<K: Key + Serialize, T: SchemaData> {
+    config: MqttConfig,
+    qos: QoS,
+    topic: String,
+    retain: bool,
+    serializer: DataSerializer<T>,
+    client: Option<AsyncClient>,
+    _t: PhantomData<K>,
+}
+
+#[process_fn(in_k = K, in_t = T)]
+impl<K: Key + Serialize, T: SchemaData + Serialize> MqttSinkFunc<K, T> {
+    pub fn new(config: MqttConfig, qos: QoS, topic: String, retain: bool, format: Format) -> Self {
+        Self {
+            config,
+            qos,
+            topic,
+            retain,
+            serializer: DataSerializer::new(format),
+            client: None,
+            _t: PhantomData,
+        }
+    }
+
+    fn name(&self) -> String {
+        format!("mqtt-producer-{}", self.topic)
+    }
+
+    pub fn from_config(config: &str) -> Self {
+        let config: OperatorConfig =
+            serde_json::from_str(config).expect("Invalid config for KafkaSink");
+        let connection: MqttConfig = serde_json::from_value(config.connection)
+            .expect("Invalid connection config for KafkaSink");
+        let table: MqttTable =
+            serde_json::from_value(config.table).expect("Invalid table config for KafkaSource");
+        let TableType::Sink { retain } = table.type_ else {
+            panic!("found non-sink mqtt config in sink operator");
+        };
+
+        Self {
+            config: connection,
+            qos: table
+                .qos
+                .and_then(|qos| match qos {
+                    QualityOfService::AtMostOnce => Some(QoS::AtMostOnce),
+                    QualityOfService::AtLeastOnce => Some(QoS::AtLeastOnce),
+                    QualityOfService::ExactlyOnce => Some(QoS::ExactlyOnce),
+                })
+                .unwrap_or(QoS::AtMostOnce),
+            topic: table.topic,
+            retain: retain
+                .and_then(|r| match r {
+                    Retain::Yes => Some(true),
+                    Retain::No => Some(false),
+                })
+                .unwrap_or_default(),
+            serializer: DataSerializer::new(
+                config.format.expect("Format must be defined for KafkaSink"),
+            ),
+            client: None,
+            _t: PhantomData,
+        }
+    }
+
+    async fn on_start(&mut self, ctx: &mut Context<(), ()>) {
+        let mut attempts = 0;
+        while attempts < 20 {
+            match super::create_connection(self.config.clone(), ctx.task_info.task_index) {
+                Ok((client, mut eventloop)) => {
+                    self.client = Some(client);
+                    tokio::spawn(async move {
```

Comment:
I changed it to not use panics, so it will reconnect instead. if for any reason a panic is thrown. The sender used by the client when publishing will fail